### PR TITLE
feat(autobump): expose more config options

### PR DESCRIPTION
**Pull request by iliapolo**
This PR exposes a few additional configuration options in order to support future `merge-back` functionality:

- `branchPrefix` - to use `merge-back` instead of `bump`.
- `branchSha` - to create the branch from `release` instead of `master`
- `labels` - to apply the `pr/no-squash` label on creation.

Note that i've currently done the minimum amount of changes so that we can move forward with the merge back PR automation. 

Its pretty apparent in my opinion that the code has become rather messy, since we are trying to support multiple scenarios using the same construct:

- Bump without a PR (for delivlib release)
- Bump with PR (for jsii and cdk)
- Bump without bump with PR (for merge back).

To that end, I have already started working on refactoring the common functionality out and making it a bit more maintainable and clear. (You can have a look here: https://github.com/awslabs/aws-delivlib/tree/epolon/mergeback)

The merge back PR automation is rather important since we are currently creating the merge-back in such a way that doesn't allow conflict resolution. I feel ok with merging this messy code and changing it when I finish the refactoring (which i am actively working on). 

Just for more context, this is how it will be used in our pipeline eventually:

```typescript
        this.pipeline.autoBump({
          bumpCommand: 'disable',
          versionCommand: `node -p "require('./scripts/get-version')"`,
          pullRequestOptions: {
            labels: ['pr/no-squash'],
            title: 'chore(release): merge release/$VERSION to master',
          },
          branchSha: 'release',
          branchPrefix: 'merge-back'
        });
```

I've also elected to currently not have the merge-back be part of the pipeline itself, but rather simply a project that will be triggered at the end of the MCM (similar to the initial bump).

I will probably integrate it as part of the pipeline along with the refactoring. In the meantime we will be _One Click Away_.

-----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lib/bump/bump.ts [hunk 1]
```diff
@@ -97,6 +97,26 @@ export interface AutoBumpOptions {
...
+  branchPrefix?: string;
+
+  /**
+   * The sha that the branch will be created from.
+   *
+   * If 'branch' is specified, this property is effectively ignored.
+   *
+   * @default 'master'
+   */
+  branchSha?: string
```
Comment (eladb) [inline lib/bump/bump.ts:118]: maybe `sourceBranch`?

File: lib/bump/bump.ts [hunk 2]
```diff
@@ -71,7 +71,7 @@ export interface AutoBumpOptions {
    * The name of the branch to push the bump commit (e.g. "master")
    * This branch has to exist.
    *
-   * @default - the commit will be pushed to the branch `bump/$VERSION`
+   * @default - the commit will be pushed to the branch `{branchPrefix}/$VERSION`
    */
   branch?: string;
```
Comment (eladb) [inline lib/bump/bump.ts:76]: Maybe rename to `targetBranch`?

File: lib/bump/bump.ts [hunk 3]
```diff
@@ -159,12 +185,15 @@ export class AutoBump extends cdk.Construct {
...
+    const branchSha = props.branchSha ?? 'master';
 
     pushCommands.push(...[
       `export VERSION=$(${versionCommand})`,
-      `export BRANCH=bump/$VERSION`,
+      `export BRANCH=${branchPrefix}/$VERSION`,
       `git branch -D $BRANCH || true`,                    // force delete the branch if it already exists
-      `git checkout -b $BRANCH`,                          // create a new branch from HEAD
+      `git checkout ${branchSha}`,                        // checkout to the sha we want to create the branch from.
+      `git checkout -b $BRANCH`,                          // create a new branch
```
Comment (eladb) [inline lib/bump/bump.ts:196]:
```suggestion
      `git checkout -b $BRANCH ${branchSha}`, // create a new branch from the source branch
```

File: lib/bump/bump.ts [hunk 4]
```diff
@@ -218,7 +247,7 @@ export class AutoBump extends cdk.Construct {
               // checked the code. Instead we define a variable that we will switch all other lines on.
               // tslint:disable-next-line:max-line-length
               `git describe --exact-match HEAD && { echo "No new commits."; export SKIP=true; } || { echo "Changes to release."; export SKIP=false; }`,
-              `$SKIP || { ${bumpCommand}; }`,
+              `$SKIP || { ${bumpCommand !== 'disable' ? bumpCommand : 'echo Bump is disabled'}; }`,
```
Comment (eladb) [inline lib/bump/bump.ts:250]: why not just `undefined` to indicate no bump command?

File: lib/bump/bump.ts [hunk 5]
```diff
@@ -272,18 +301,26 @@ function createPullRequestCommands(repo: WritableGitHubRepo, options: PullReques
...
   ];
 
+  if ((options.labels ?? []).length !== 0) {
+    commands.push(`${curl(`/issues/$PR_NUMBER/labels`, '-X POST', {'labels': options.labels}, repo)}`);
+  }
+
+  return commands;
+}
+
+function curl(uri: string, command: string, request: any, repo: WritableGitHubRepo): string {
```
Comment (eladb) [inline lib/bump/bump.ts:317]: rename to `githubCurl`

File: lib/bump/bump.ts [hunk 6]
```diff
@@ -272,18 +301,26 @@ function createPullRequestCommands(repo: WritableGitHubRepo, options: PullReques
...
-    `-X POST`,
-    `--header "Authorization: token $GITHUB_TOKEN"`,
-    `--header "Content-Type: application/json"`,
-    `-d ${JSON.stringify(JSON.stringify(request))}`, // to escape quotes
-    `https://api.github.com/repos/${repo.owner}/${repo.repo}/pulls`
+  const commands = [
+     condition,
+    `GITHUB_TOKEN=$(aws secretsmanager get-secret-value --secret-id "${repo.tokenSecretArn}" --output=text --query=SecretString)`,
+    `${curl('/pulls', '-X POST -o pr.json', request, repo)} && export PR_NUMBER=$(node -p 'require("./pr.json").number')`
   ];
```
Comment (NetaNir) [inline lib/bump/bump.ts:308]: Why write the PR numbers to a file and also export it? The commend that requires the issue number can read it from the file (same way at the export is created)

File: lib/bump/bump.ts [hunk 7]
```diff
@@ -127,6 +147,12 @@ export interface PullRequestOptions {
    * @default "master"
    */
   base?: string;
+
+  /**
+   * Set of labels to apply to the PR.
+   */
+  labels?: string[]
+
```
Comment (NetaNir) [inline lib/bump/bump.ts:155]: `Labels` should be added on type `PullRequestOptions`  which is included in `AutoBumpOptions`

File: lib/bump/bump.ts (hunk 1 above)
Comment (iliapolo) [reply lib/bump/bump.ts:118]: 👍

File: lib/bump/bump.ts (hunk 2 above)
Comment (iliapolo) [reply lib/bump/bump.ts:76]: Maybe, but not worth the breaking change for now. Its going to look different once I finish the refactoring anyway, so lets leave it for now.

File: lib/bump/bump.ts (hunk 3 above)
Comment (iliapolo) [reply lib/bump/bump.ts:196]: 👍

File: lib/bump/bump.ts (hunk 4 above)
Comment (iliapolo) [reply lib/bump/bump.ts:250]: I'd love to, but there is a default: `/bin/bash ./bump.sh`.

We had the same problem with `scheduleExpression`:

https://github.com/awslabs/aws-delivlib/blob/a0168d7cdd15dbb5d51bc1a8236048c873906fc9/lib/bump/bump.ts#L243-L253

Part of the reason I want to refactor...

File: lib/bump/bump.ts (hunk 5 above)
Comment (iliapolo) [reply lib/bump/bump.ts:317]: 👍

File: lib/bump/bump.ts (hunk 7 above)
Comment (iliapolo) [reply lib/bump/bump.ts:155]: Its is in `PullRequestOptions` - am I missing something?

File: lib/bump/bump.ts (hunk 6 above)
Comment (iliapolo) [reply lib/bump/bump.ts:308]: We might have additional commands that need access to the PR number. This way users can also just use `$PR_NUMBER` in their own custom PR body.

File: lib/bump/bump.ts (hunk 7 above)
Comment (NetaNir) [reply lib/bump/bump.ts:160]: dammit, the diff view collapsed the `export interface PullRequestOptions` from the view  so it looked like it was added on `AutoBumpProps`, I will see myself out.

**Pull request merged**